### PR TITLE
Support read-only sessions

### PR DESCRIPTION
**Pull request by amolenaar**
# Description

* A button to start a new read-only session is placed in the model card
* The button will load all models that match the project and tool (version) of the loaded model

NB. Read-only models only apply to models available from a git repository.

Resolves #issue

# Testing

How were the changes tested?

# Checklist

- [x] My code follows the guidelines of this project: [CONTRIBUTING.md](/CONTRIBUTING.md)
- [x] I updated the documentation with the new functionality
- [x] I went through the code and removed all print statements, breakpoints and unused code
- [x] Error handling for all possible scenarios has been implemented
- [x] I've add logging for all relevant operations
- [x] I added the appropriate labels to this PR (enhancement/bug/documentation, effort, priority)

# Current state (13.10.2022)

* Backend is working with limitations. It works only with one model: the first model that matches the requested tool/version.
* The frontend is working with fixed data: project is "default" and the tool is Capella 5.2.

What needs to be done still:

- [x] Update the "readonly" docker image to support multiple repositories
- [x] It's now clear that, when the "create session" button is hit something's happening. It needs some feedback
- [x] With me, the state keeps hanging in a "prepare" state. Not sure what happens there.

### CODE REVIEW

File: backend/alembic_revision.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        os.environ["ALEMBIC_SQLALCHEMY_URL"] = database_url
+
+        alembic_cfg = Config("./alembic.ini")
+        command.revision(alembic_cfg, message=summary, autogenerate=True)
+
+
+if __name__ == "__main__":
+    logging.basicConfig(level=logging.DEBUG)
+
+    main(summary=sys.argv[1])
```
Comment (MoritzWeber0) [inline backend/alembic_revision.py:59]: I don't get the point of this script. Alembic compares the current database state with how it should look like (by looking to the code) when autogenerating the revision. When you spawn a database before and initialize it with the state from the code, then there is exactly no difference between the current database structure and the code - and alembic doesn't detect the changes.

My workflow to run the migrations is to setup a database, initialize it with the current state and then make the changes to the code. When I'm done with it, I run `alembic revision --autogenerate`. Because my database structure has the old version, alembic picks up the changes.

File: backend/capellacollab/sessions/routes.py [hunk 2]
```diff
@@ -87,70 +96,66 @@ def get_current_sessions(
...
+            status_code=404,
             detail={
-                "err_code": "wrong_depth_format",
-                "reason": f"Depth type {depth} is not allowed.",
+                "err_code": "git_model_not_found",
+                "reason": "The Project has no connected Git Model. Please contact a project manager or admininistrator",
             },
         )
+
+    docker_image = get_readonly_image_for_version(model.version)
```
Comment (TimoBbz) [inline backend/capellacollab/sessions/routes.py:153]: When the `version.tool` has no `readonly_docker_image_template`, this call raises an error that should be handled (or even better imo, the button should not be availble when there is no `readonly_docker_image_template`).

File: backend/capellacollab/tools/models.py [hunk 3]
```diff
@@ -1,4 +1,5 @@
 # SPDX-FileCopyrightText: Copyright DB Netz AG and the capella-collab-manager contributors
```
Comment (TimoBbz) [inline backend/capellacollab/tools/models.py:1]: There is one license line too many.
Also, the `ToolDockerimage` currently uses the `docker_image_template` as readonly (in `from_orm`, that is used under the hood everytime a FastAPI route has a response_model set.), I don’t think that’s the desired behaviour.

**Comment by TimoBbz** [comment]
The editing of the readonly docker image template is not handled.

File: backend/capellacollab/sessions/routes.py [hunk 4]
```diff
@@ -159,10 +164,35 @@ def request_session(
         owner,
         rdp_password,
         db,
-        repository=body.repository,
+        repository=[],
```
Comment (TimoBbz) [inline backend/capellacollab/sessions/routes.py:167]: This parameter doesn’t accept a string, either the `create_database_and_guacamole_session` function has not been updated or the repository should be left unset.

File: frontend/src/app/helpers/skeleton-loaders/mat-card-overview-loader/mat-card-overview-loader.component.spec.ts [hunk 5]
```diff
@@ -1,27 +0,0 @@
-/*
```
Comment (TimoBbz) [inline frontend/src/app/helpers/skeleton-loaders/mat-card-overview-loader/mat-card-overview-loader.component.spec.ts:1]: Why was this file removed?

File: frontend/src/app/services/model/model.service.ts [hunk 6]
```diff
@@ -28,8 +28,8 @@ export interface Model {
   tool: Tool;
   version: ToolVersion | null;
   type: ToolType | null;
-  t4c_model_id: number | null;
-  git_model_id: number | null;
+  t4c_models: Object | null;
+  git_models: Object | null;
```
Comment (TimoBbz) [inline frontend/src/app/services/model/model.service.ts:32]: You should specify the type of `t4c_models` and `git_models`, and also rather use the interrogative syntax.

```typescript
version?: ToolVersion;
type?: ToolType;
t4c_models?: T4CModel[];
git_models?: GitModel[];
```

File: frontend/src/app/settings/modelsources/t4c-settings/edit-t4c-instance/edit-t4c-instance.component.spec.ts [hunk 7]
```diff
@@ -1,27 +0,0 @@
-/*
```
Comment (TimoBbz) [inline frontend/src/app/settings/modelsources/t4c-settings/edit-t4c-instance/edit-t4c-instance.component.spec.ts:1]: Why removing this file?

File: frontend/src/app/settings/modelsources/t4c-settings/t4c-settings-wrapper/t4c-settings-wrapper.component.spec.ts [hunk 8]
```diff
@@ -1,27 +0,0 @@
-/*
```
Comment (TimoBbz) [inline frontend/src/app/settings/modelsources/t4c-settings/t4c-settings-wrapper/t4c-settings-wrapper.component.spec.ts:1]: Same question

File: frontend/src/app/projects/project-detail/model-overview/model-overview.component.ts [hunk 9]
```diff
@@ -31,12 +36,23 @@ export class ModelOverviewComponent implements OnInit {
...
       return 'T4C';
-    } else if (model.git_model_id) {
+    } else if (model.git_models) {
       return 'Git';
     } else {
       return 'Unset';
     }
   }
+
+  requestSession(model: Model) {
```
Comment (TimoBbz) [inline frontend/src/app/projects/project-detail/model-overview/model-overview.component.ts:48]: You should specify the type signature (`void`) of this function.

File: frontend/src/app/projects/project-detail/model-overview/model-overview.component.ts [hunk 10]
```diff
@@ -4,11 +4,14 @@
...
 
 import { Component, Input, OnInit } from '@angular/core';
+import { Router } from '@angular/router';
+
 import { Model, ModelService } from 'src/app/services/model/model.service';
 import {
   Project,
   ProjectService,
 } from 'src/app/services/project/project.service';
+import { SessionService } from '../../../services/session/session.service';
```
Comment (TimoBbz) [inline frontend/src/app/projects/project-detail/model-overview/model-overview.component.ts:14]: Prefer absolute (from `src`) paths for imports.

File: backend/capellacollab/sessions/routes.py [hunk 11]
```diff
@@ -87,70 +96,66 @@ def get_current_sessions(
...
+                "technical": f"No project {body.project_slug} found.",
             },
         )
-    verify_project_role(repository=body.repository, token=token, db=db)
-    git_model = git_models_crud.get_primary_model_of_repository(
-        db, body.repository
-    )
-    if not git_model:
+
+    verify_project_role(project.name, token, db)
```
Comment (TimoBbz) [inline backend/capellacollab/sessions/routes.py:127]: If you rebased on _v2/staging_, you should rather use the injectables to check the permissions.

File: backend/capellacollab/sessions/routes.py [hunk 12]
```diff
@@ -87,70 +96,66 @@ def get_current_sessions(
...
-    assert body.type == WorkspaceType.READONLY
-
     rdp_password = generate_password(length=64)
 
-    owner = db_user.username
+    owner = db_user.name
 
     log.info("Starting persistent session creation for user %s", owner)
 
     existing_user_sessions = database.get_sessions_for_user(db, owner)
```
Comment (TimoBbz) [inline backend/capellacollab/sessions/routes.py:115]: This variable is no more used.

**Comment by TimoBbz** [comment]
A general comment, I think in this case there should be no alembic merge since there is no need to support migrations of the diverging histories (since there is no production environment that currently uses the intermediate alembic commits of v2). I’d advise for a rebase, then for a manual chaining of the migrations without merges.

File: frontend/src/app/helpers/skeleton-loaders/mat-card-overview-loader/mat-card-overview-loader.component.spec.ts (hunk 5 above)
Comment (amolenaar) [reply frontend/src/app/helpers/skeleton-loaders/mat-card-overview-loader/mat-card-overview-loader.component.spec.ts:1]: I think I thought it was no longer needed, but I see it referenced in the project overview still.

Comment (amolenaar) [reply frontend/src/app/helpers/skeleton-loaders/mat-card-overview-loader/mat-card-overview-loader.component.spec.ts:1]: restored.

File: frontend/src/app/services/model/model.service.ts (hunk 6 above)
Comment (amolenaar) [reply frontend/src/app/services/model/model.service.ts:32]: Type is probably indeed. Thanks for making me aware.

File: frontend/src/app/settings/modelsources/t4c-settings/edit-t4c-instance/edit-t4c-instance.component.spec.ts (hunk 7 above)
Comment (amolenaar) [reply frontend/src/app/settings/modelsources/t4c-settings/edit-t4c-instance/edit-t4c-instance.component.spec.ts:1]: I'm not sure. I removed some files for my previous PR (persistent sessions), but not for this PR as far as I can remember.

Comment (amolenaar) [reply frontend/src/app/settings/modelsources/t4c-settings/edit-t4c-instance/edit-t4c-instance.component.spec.ts]: I undid the removal.

File: frontend/src/app/services/model/model.service.ts (hunk 6 above)
Comment (TimoBbz) [reply frontend/src/app/services/model/model.service.ts]: You should anyways rebase on v2/staging, because the `t4c_models` and `git_models` properties are handled anyways tho.

**Comment by amolenaar** [comment]
> I’d advise for a rebase, then for a manual chaining of the migrations without merges.

Not again :see_no_evil: !

File: frontend/src/app/projects/project-detail/model-overview/model-overview.component.ts (hunk 10 above)
Comment (amolenaar) [reply frontend/src/app/projects/project-detail/model-overview/model-overview.component.ts:14]: :+1:

**Comment by TimoBbz** [comment]
When Guacamole is unreachable, some `GuacamoleError` (⁉️) are raised. This should be excepted or replaced with `fastapi.HTTPException` at some point.

File: backend/capellacollab/projects/capellamodels/models.py [hunk 13]
```diff
@@ -76,8 +84,12 @@ class DatabaseCapellaModel(Base):
 
     editing_mode = Column(Enum(EditingMode))
 
-    t4c_models = relationship("DatabaseT4CModel", back_populates="model")
-    git_models = relationship("DB_GitModel", back_populates="model")
+    t4c_models: t.Sequence[DatabaseT4CModel] = relationship(
```
Comment (TimoBbz) [inline backend/capellacollab/projects/capellamodels/models.py:87]: I cannot find any documentation on the type returned by the relationship, how do you know it’s a `Sequence`?

Comment (TimoBbz) [reply backend/capellacollab/projects/capellamodels/models.py:87]: I finally found that it’s an [InstrumentedList](https://docs.sqlalchemy.org/en/20/orm/collection_api.html#sqlalchemy.orm.collections.InstrumentedList), which inherits from the list.
I don’t know what to do about this information tho 🙊

File: backend/capellacollab/sessions/routes.py (hunk 11 above)
Comment (amolenaar) [reply backend/capellacollab/sessions/routes.py:129]: I tried that, but it made my tests fail. I'd like to leave it as is for now, and have a look at improving the code in this area later.

File: backend/capellacollab/projects/capellamodels/models.py (hunk 13 above)
Comment (amolenaar) [reply backend/capellacollab/projects/capellamodels/models.py:87]: A sequence is a primitive type for a list, so that should be fine. We do not need the complete list interface so far.

File: backend/capellacollab/sessions/routes.py (hunk 2 above)
Comment (amolenaar) [reply backend/capellacollab/sessions/routes.py:155]: For now, the backend will return a 409 (Conflict) to address the missing read-only docker image.

File: backend/capellacollab/tools/models.py [hunk 14]
```diff
@@ -26,6 +26,7 @@ class Tool(Base):
     id = Column(Integer, primary_key=True)
     name = Column(String)
     docker_image_template = Column(String)
+    readonly_docker_image_template = Column(String)
```
Comment (TimoBbz) [inline backend/capellacollab/tools/models.py:29]: The Pydantic model `ToolDockerImage` should also be updated to return `obj.readonly_docker_image_template` in the `from_orm` method.

Comment (TimoBbz) [reply backend/capellacollab/tools/models.py:29]: Also, the update route should be fixed.

File: frontend/src/app/projects/project-detail/model-overview/model-overview.component.html [hunk 15]
```diff
@@ -4,12 +4,12 @@
  -->
 
 <div class="models" *ngIf="models">
-  <div class="models-title">
-    <h2 id="title">Models</h2>
+  <h2>
+    Models
```
Comment (MoritzWeber0) [inline frontend/src/app/projects/project-detail/model-overview/model-overview.component.html:8]: The new design on broken on my devices (the old as well). My propsal

HTML:
```
  <div class="models-title">
    <h2 id="title">Models</h2>
    <a mat-button [routerLink]="['/project', project.slug, 'models', 'create']">
      <mat-icon id="add-icon">add</mat-icon>
    </a>
  </div>
```

CSS:
```
.models-title {
  display: flex;
  align-items: center;
}

#title {
  margin: 0;
}
```

and remove the `#title` rule.

File: backend/capellacollab/sessions/routes.py [hunk 16]
```diff
@@ -87,70 +100,72 @@ def get_current_sessions(
...
             },
         )
 
-    revision = body.branch or git_model.revision
-    if body.depth == DepthType.LatestCommit:
-        depth = 1
-    elif body.depth == DepthType.CompleteHistory:
-        depth = 0
-    else:
+    docker_image = get_readonly_image_for_version(model.version)
```
Comment (MoritzWeber0) [inline backend/capellacollab/sessions/routes.py:155]: I'm not able to set the `readonly` image. To reproduce: 
Navigate to http://localhost:4200/settings/core/tool/1, change the read-only image and refresh the page. It shows the default again. 
I couldn't implement it due to the missing template in the database, but it should be an easy fix. Check the `FIXME` in capellacollab/tools/crud.py

**Review by MoritzWeber0 — CHANGES_REQUESTED** [review]
Thanks for the improvement @amolenaar and thanks @TimoBbz for the review. I have some additional change requests.

**Comment by MoritzWeber0** [comment]
> When Guacamole is unreachable, some `GuacamoleError` (⁉️) are raised. This should be excepted or replaced with `fastapi.HTTPException` at some point.

Nope. This is nothing that a user should see (as the installation itself is broken then) and you need to contact an administrator anyway.

File: backend/capellacollab/tools/models.py (hunk 14 above)
Comment (amolenaar) [reply backend/capellacollab/tools/models.py:29]: fixed.

File: backend/capellacollab/alembic/versions/ab01ad045341_add_readonly_docker_image.py [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+depends_on = None
+
+
+def upgrade():
+    with op.batch_alter_table("tools") as batch_op:
+        batch_op.add_column(
+            sa.Column(
+                "readonly_docker_image_template", sa.String(), nullable=True
+            )
+        )
```
[the remark this answers is not in the document]
Comment (amolenaar) [reply backend/capellacollab/alembic/versions/ab01ad045341_add_readonly_docker_image.py]: fixed

File: backend/capellacollab/projects/capellamodels/models.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (amolenaar) [reply backend/capellacollab/projects/capellamodels/models.py:87]: Didn't know ;) fixed.

File: backend/capellacollab/sessions/routes.py [hunk 18]
```diff
@@ -159,10 +174,34 @@ def request_session(
...
+        yield git_model_as_json(git_model)
+
+
+def git_model_as_json(git_model: DB_GitModel) -> dict[str, str | int]:
+    json = {
+        "url": git_model.path,
+        "revision": git_model.revision,
+        "depth": 1,
+        "entrypoint": git_model.entrypoint,
+        "nature": "project",
```
[the remark this answers is not in the document]
Comment (amolenaar) [reply backend/capellacollab/sessions/routes.py]: fixed

File: backend/capellacollab/sessions/routes.py [hunk 19]
```diff
@@ -159,10 +174,34 @@ def request_session(
         owner,
         rdp_password,
         db,
-        repository=body.repository,
```
[the remark this answers is not in the document]
Comment (amolenaar) [reply backend/capellacollab/sessions/routes.py:162]: Fixing this is a PR in itself :smile:

File: backend/capellacollab/sessions/routes.py (hunk 16 above)
Comment (amolenaar) [reply backend/capellacollab/sessions/routes.py:152]: I tried and it seems fixed now.

File: frontend/src/app/projects/project-detail/model-overview/model-overview.component.html (hunk 15 above)
Comment (amolenaar) [reply frontend/src/app/projects/project-detail/model-overview/model-overview.component.html]: I do not see that problem, but added the extra div.

File: backend/capellacollab/alembic/versions/ab01ad045341_add_readonly_docker_image.py [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+    op.add_column(
+        "tools",
+        sa.Column(
+            "readonly_docker_image_template", sa.String(), nullable=True
+        ),
+    )
+
+
+def downgrade():
+    with op.batch_alter_table("tools") as batch_op:
```
Comment (MoritzWeber0) [inline backend/capellacollab/alembic/versions/ab01ad045341_add_readonly_docker_image.py:31]: Still one old `batch_alter_table` that can be replaced  :)

Comment (amolenaar) [reply backend/capellacollab/alembic/versions/ab01ad045341_add_readonly_docker_image.py]: :see_no_evil:

File: backend/capellacollab/sessions/routes.py [hunk 21]
```diff
@@ -159,10 +174,34 @@ def request_session(
...
 
+def models_as_json(models: t.List[DatabaseCapellaModel], version: Version):
+    for m in models:
+        if m.version.id == version.id:
+            yield from model_as_json(m)
+
+
+def model_as_json(model: DatabaseCapellaModel):
+    for git_model in model.git_models:
+        yield git_model_as_json(git_model)
```
[the remark this answers is not in the document]
Comment (amolenaar) [reply backend/capellacollab/sessions/routes.py:188]: I prefer a few smaller methods. Otherwise the nesting will be really deep. The intent is to let the function names express the intent.

File: backend/capellacollab/sessions/routes.py [hunk 22]
```diff
@@ -87,70 +100,72 @@ def get_current_sessions(
...
-            status_code=404,
-            detail={
-                "err_code": "existing_session",
-                "reason": f"You already have a open Read-Only Session for the repository {body.repository}. Please navigate to 'Active Sessions' to Reconnect",
+            status.HTTP_404_NOT_FOUND,
+            {
+                "reason": f"The model with name {body.model_slug} was not found.",
+                "technical": f"No model {body.model_slug} found.",
             },
         )
```
[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply backend/capellacollab/sessions/routes.py]: How is this resolved? There is no comment and I don't see it applied.

[the remark this answers is not in the document]
Comment (amolenaar) [reply backend/capellacollab/sessions/routes.py]: Yes it is: the comment is marked "outdated"

[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply backend/capellacollab/sessions/routes.py]: It shows as outdated, but I don't see the suggestion applied. Am I missing something?

[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply backend/capellacollab/sessions/routes.py]: image

File: backend/capellacollab/sessions/routes.py [hunk 23]
```diff
@@ -159,10 +174,34 @@ def request_session(
...
         rdp_password,
         db,
-        repository=body.repository,
     )
 
 
+def models_as_json(models: t.List[DatabaseCapellaModel], version: Version):
+    for m in models:
+        if m.version.id == version.id:
+            yield from model_as_json(m)
```
[the remark this answers is not in the document]
Comment (amolenaar) [reply backend/capellacollab/sessions/routes.py:183]: True. Filtering in the route was added later. So this can be simplified.

File: backend/capellacollab/sessions/routes.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (amolenaar) [reply backend/capellacollab/sessions/routes.py]: > Same here. Please use the `injectables` instead: `get_existing_capella_model`

This does not apply, since the `model_slug` is not part of the URL. I can reuse that function though.

**Review by MoritzWeber0 — APPROVED** [review]
